src/hdf5_dl.c

```diff
@@ -97,44 +105,204 @@ hid_t H5E_ERR_CLS_g = -1;
 
 static bool is_init = false;
 
+/*
+ * Try to find a symbol within a library
+ * 
+ * handle: Handle to the library
+ * symbol: Symbol to look for
+ * Returns: a pointer to the symbol or NULL 
+ * if the symbol can't be found 
+ */
+void *find_sym(void *handle, const char *symbol) {
+
+  void *ret = NULL, *err = NULL;
+  dlerror(); /* clear error code */
+  ret = dlsym(handle, symbol);
+
+  if(ret != NULL && (err = dlerror()) == NULL)
+    return ret;
+  else
+    return NULL;
+}
+
+/*
+ * Check that all symbols have been loaded
+ * 
+ * Returns: -1 if an error occured, 0 for success
+ */
+int check_symbols() {
+
+  if(DL_H5Functions.H5open == NULL)
+    return -1;
+
+  /*H5E*/
+  if(DL_H5Functions.H5Epush1 == NULL)
+    return -1;
+
+  if(DL_H5Functions.H5Epush2 == NULL)
+    return -1;
+
+  /*H5P*/
+  if(DL_H5Functions.H5Pget_filter_by_id2 == NULL)
+    return -1;
+
+  if(DL_H5Functions.H5Pget_chunk == NULL)
+    return -1;
+
+  if(DL_H5Functions.H5Pmodify_filter == NULL)
+    return -1;
+
+  /*H5T*/
+  if(DL_H5Functions.H5Tget_size == NULL)
+    return -1;
+
+  if(DL_H5Functions.H5Tget_class == NULL)
+    return -1;
+  
+  if(DL_H5Functions.H5Tget_super == NULL)
+    return -1;
+  
+  if(DL_H5Functions.H5Tclose == NULL)
+    return -1;
+
+  /*H5Z*/
+  if(DL_H5Functions.H5Zregister == NULL)
+    return -1;
+
+  /*Variables*/
+  if(H5Variables_ptr.h5e_cantregister_ptr == NULL)
+    return -1;
+
+  if(H5Variables_ptr.h5e_callback_ptr == NULL)
+    return -1;
+
+  if(H5Variables_ptr.h5e_pline_ptr == NULL)
+    return -1;
+
+  if(H5Variables_ptr.h5e_err_cls_ptr == NULL)
+    return -1;
+
+  return 0;
+
+}
 
 /* Initialize the dynamic loading of symbols and register the plugin
  *
  * libname: Name of the DLL from which to load libHDF5 symbols
- * Returns: a value < 0 if an error occured
+ * Returns: -1 if an error occured, 0 for success
  */
-int init_filter(const char* libname)
+int init_filter(const char *libname)
 {
     int retval = -1;
-  	void * handle;
+  	void *handle = NULL, *symbol = NULL;
 
     handle = dlopen(libname, RTLD_LAZY | RTLD_LOCAL);
 
     if (handle != NULL) {
         /*H5*/
-        DL_H5Functions.H5open = (DL_func_H5open)dlsym(handle, "H5open");
+        if(DL_H5Functions.H5open == NULL) {
+            symbol = find_sym(handle, "H5open");
+            if(symbol != NULL)
+                DL_H5Functions.H5open = (DL_func_H5open)symbol;
+        }
+
         /*H5E*/
-        DL_H5Functions.H5Epush1 = (DL_func_H5Epush1)dlsym(handle, "H5Epush1");
-        DL_H5Functions.H5Epush2 = (DL_func_H5Epush2)dlsym(handle, "H5Epush2");
+        if(DL_H5Functions.H5Epush1 == NULL) {
+            symbol = find_sym(handle, "H5Epush1");
+            if(symbol != NULL)
+                DL_H5Functions.H5Epush1 = (DL_func_H5Epush1)symbol;
+        }
```

Comment:
You could consider some ways to shorten this boiler plate:
- Wrap it up in a function or macro
- Don't bother with the conditional assignment as it doesn't do anything (i.e. you know the struct entry is NULL, so when the symbol is NULL, assigning it to the entry does nothing). Maybe leave a comment to that effect on the first one as not checking the value could look like a bug.

TBH if you do the latter, there's probably not enough boilerplate left to warrant the former.

Reply:
Fixed this now.